lib/LaTeXML/Package/a11ymark.sty.ltxml

```diff
@@ -0,0 +1,93 @@
+# |---------------------------------------------------------------------| #
+# | Bruce Miller <bruce.miller@nist.gov>                        #_#     | #
+# | http://dlmf.nist.gov/LaTeXML/                              (o o)    | #
+# \=========================================================ooo==U==ooo=/ #
+package LaTeXML::Package::Pool;
+use strict;
+use warnings;
+use LaTeXML::Package;
+
+## 0. Semantic infrastructure
+DefConstructor('\meaning{}{}', '#2', afterConstruct => sub {
+    my ($doc, $whatsit) = @_;
+    my $node  = $doc->getNode;
+    my $apply = $node->lastChild;
+    if ($apply && $apply->localName eq 'XMApp') {    # should be an apply, if by design
+      $apply->setAttribute('meaning', ToString($whatsit->getArg(1))); }
+    else {    # otherwise, add the meaning on the current node, possibly an empty XMArg
+      $node->setAttribute('meaning', ToString($whatsit->getArg(1))); }
+    return; }, bounded => 1, requireMath => 1);
+
+# Embellishment is hard to write, hard to speak, but describes exactly several cases
+# I will abbreviate it "emb", for now, and use it as a prefix
+DefConstructor('\emb@atom{}{}',
+  '<ltx:XMDual>'
+    . '<ltx:XMTok meaning="#1" />'
+    . '<ltx:XMWrap>#2</ltx:XMWrap>'
+    . '</ltx:XMDual>',
+  bounded => 1, requireMath => 1);
+
+DefMacro('\emb@base{}{}{}', sub {
+    my ($gullet, $meaning, $base, $emb) = @_;
+    # Package::dualize_arglist seems to be doing this with more sophistication,
+    # a good bit to read if we ever need to build more infra here.
+    # for now, fast and loose, these don't get saved and are only for local association
+    my $id_token  = T_OTHER("id" . int(rand(10000)));
+    my $ref_token = Invocation(T_CS('\@XMRef'), $id_token);
+    my $arg_token = Invocation(T_CS('\@XMArg'), $id_token, $base);
+    return Invocation(T_CS('\emb@base@build'),
+      $meaning, $ref_token, Tokens($arg_token, $emb))->unlist });
+
+DefMacro('\numprimemarks{}', sub {
+    my ($gullet, $numtok) = @_;
+    my $num = int(ToString($numtok));
+    return map { T_CS('\prime') } 1 .. $num; });
+
+DefConstructor('\emb@base@build{}{}{}',
+  '<ltx:XMDual>'
+    . '<ltx:XMApp>'
+    . '<ltx:XMTok meaning="#1"/>'
+    . '#2'
+    . '</ltx:XMApp>'
+    . '<ltx:XMWrap>#3</ltx:XMWrap>'
+    . '</ltx:XMDual>',
+  bounded => 1, requireMath => 1);
+
+## I. Calculus
+DefConstructor('\diffd', '<ltx:XMTok meaning="differential" name="diffd" role="DIFFOP">d</ltx:XMTok>');
+DefMath('\deriv[]{}{}',
+  '\frac{\@MAYBEAPPLY{\@SUPERSCRIPT{\diffd}{#1}}{#2}}'
+    . '{\@SUPERSCRIPT{\@APPLY{\diffd #3}}{#1}}',
+  meaning => 'derivative', reorder => [2, 3, 1],
+  # afterDigest => sub {
+  #   # NOTE: arg 2 will be wrapped in XMRef!
+  #   $_[1]->setProperty(role => 'DIFFOP') if checkDiffOp($_[1]);
+  #   return; },
+  hide_content_reversion => 1);
+
+DefMath('\integral{}{}', '\int #1 \diffd #2', meaning => 'integral');
+
+## II. Scripts
+DefMacro('\power{}{}',   '\meaning{power}{#1^{#2}}');
```

Comment:
Try
```
DefMath('\power{}{}', "{#1^{#2}}", meaning => 'power',
  reversion              => '#1^{#2}',
  hide_content_reversion => 1);
```

Reply:
It works too it seems! Yours creates the markup:
```xml
  <XMath>
    <XMDual>
      <XMApp>
        <XMTok meaning="power" role="UNKNOWN"/>
        <XMRef idref="p1.m1.1"/>
        <XMRef idref="p1.m1.2"/>
      </XMApp>
      <XMApp>
        <XMTok role="SUPERSCRIPTOP" scriptpos="post2"/>
        <XMTok role="UNKNOWN" xml:id="p1.m1.1" font="italic" fragid="p1.m1.1">x</XMTok>
        <XMTok meaning="2" role="NUMBER" xml:id="p1.m1.2" fontsize="70%" fragid="p1.m1.2">2</XMTok>
      </XMApp>
    </XMDual>
  </XMath>
```

my `\meaning` take does:
```xml
  <XMath>
    <XMApp>
      <XMTok meaning="power" role="SUPERSCRIPTOP" scriptpos="post2"/>
      <XMTok role="UNKNOWN" font="italic">x</XMTok>
      <XMTok meaning="2" role="NUMBER" fontsize="70%">2</XMTok>
    </XMApp>
  </XMath>
```

which improves the tree but loses some attributes. I wonder if I should be modding `DefMath` instead of creating more magical  `DefConstructor`s...

Reply:
Don't know about "improves", depends on what you like :>  Your `\meaning` probably works in most simple cases, but (you imply) needs extra mods.  `DefMath` is exactly that magical DefConstructor that already knows about the arguments, so it tends to put the attributes (meaning) in the right place.  With the side benefit that it creates an XMDual! :>

Reply:
Is this really down to personal preference in these two examples? I thought it was an XMath design goal to capture the `meaning` attribute of terminal tokens where they are clear. Would you say it's personal preference if the `meaning="2"` on the number token sits in a dual, as in:
```xml
<XMDual>
  <XMTok meaning="2" role="UNKNOWN"/>
  <XMTok role="NUMBER" xml:id="p1.m1.2" fontsize="70%" fragid="p1.m1.2">2</XMTok>
</XMDual>
```

Such tokens are also not cross-referenced, so the connection is less direct than it would be with an `XMRef`. I was approaching things with the mindset that the duals ought to only come into play where the tree correspondence between presentation and content breaks down, if we're to think of some notion of "optimal" XMath.

Reply:
To put it simply, I'm fishing for a `DefMath` form which first instantiates the XMath tree as usual, then picks the operating token and assigns it a meaning. Maybe a new option for DefMath that insists on marking up in-place... (or an existing one, I'm still getting the hang of the advanced DefMath uses)

Reply:
I think I've had this debate with myself at least as many times as I've had it with you :>

There's Taste, and there's the Possible, and then there are the Traps :>  `DefMath` tries hard to do the right thing, and has been through many iterations of traps, so you want to think twice before reinventing it. OTOH, it's a bit conservative.  So, for example, the `\diffd` case decided on a dual because it saw `\mathrm` and got suspicious.  A better definition would be
```
DefMath('\diffd', 'd',
  meaning => 'differential', role => 'DIFFOP',
  font    => { shape => 'upright' });
```
which avoids the dual.

Next up is the "clever" way that `role` is used that allows a simple triple to represent both the presentation and content trees in many, but not all, cases. Alas, again `DefMath` isn't quite smart enough to construct a sample and figure out that the presentation & content trees are isomorphic, and thus avoid a dual.  Basically, it sees that there are arguments and punts to a dual to be sure. Perhaps that's a further enhancement? Or lacking that, an extra option that asserts what kind of construction should be used?

The trap with a simple version, like `\meaning`, is that you grow to like it and start using it all over and then realize it's doing weird things, so it grows and handles various special cases until it ends up being... `DefMath`!

Another trap, and something I struggled with a few commits back, was wanting to immediately optimize some special cases that *should* produce something more canonical, but realized that the remaining machinery needed to handle the awkward markup as well as the simpler. So I put off that optimization for a bit.

So, maybe we could make `\power` cleaner, but not till after you make the annotation work with the bad markup, first :>

Reply:
Thanks, I quite like this poetic overview of Taste, The Possible and Traps :> It even feels close to a documentation/writeup heading, just needs a third t-word. "Tools, Taste and Traps"?

You've successfully convinced me I should be using DefMath! But your comments in DefMath had convinced me that you've had my problems before and have been hitting similar walls namely:
```
# OK, this is getting a bit out-of-hand; I can't, myself, predict whether XMDual gets involved!
```

So I'll try to abolish my DefConstructor games with `\meaning`, and instead try to use DefMath exclusively, making it "less conservative" where needed - as you suggest. I think I only have a couple of new cases that I'm interested in (not making a dual when a script has a clear meaning but using a compact XMApp, maybe a general rewrite rule that compacts redundant duals into apps...).

Since the DefMath interface allows for open-ended key=>val options, I can just hide whatever I need behind a new option name, keep the defaults nice and conservative :>  Thanks for the great feedback!